text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+## Use `uint<N>` and `int<N>` instead of `u<N>` and `i<N>`
+
+This was the previous proposal used by the original RFC, and they were chosen due to a misconception about name conflicts. Although the `i` variable is commonly used for loops and other index variables, variable names and type names are entirely separate and can coexist peacefully. Having names closer to the originals feels more discoverable here.
+
+Additionally, the names `uint` and `int` remain available for future, potentially unrelated integer types.
+
+## Bound-based generalisation
+
+Generalising integers over their number of bits is certainly a very logical way to generalise integers *for computers*, but generalising based upon bounds is a very natural way for humans to do it, *and* more general. For example, instead of `u<N>` and `i<N>` types, we could get away with just one type,
+`int<A..=B>`. This would be more powerful than the original: for example, a percentage could be represented exactly as `int<0..=100>`. Whether an integer is signed simply depends on whether its lower bound is negative.
+
+The primary reason for leaving this out is… well, it's a lot different from the existing integer types in the language. Additionally, such a proposal can coexist with this one, since there's nothing stopping us from making an `int<A..=B>` type in the future such that `u<N>` and `i<N>` are equivalent to the correct range of integers.
+
+## Integer traits
+
+Previously, Rust had traits which generalized the integer types and their methods, but these were ultimately removed in favour of inherent methods. Going with a generic `u<N>` over an `Int` trait would avoid the problem of determining which methods are suited for these traits; instead, the `u<N>` type would have all of them.
+
+Additionally, having these traits does not allow non-power-of-two `u<N>` outright, and so, this feature is a strict superset of that.
+
+Having separate `u<N>` and `i<N>` types also solves the problem of generalising signed and unsigned integers: everything would require one impl for signed integers, and one for unsigned. This would ensure that these types have exactly the same behaviour for all `N`, only differing in the upper bound on the number of bits.
+
+## Offering as a library
+
+This was the main proposal last time this RFC rolled around, and as we've seen, it hasn't really worked.
+
+Crates like [`u`], [`bounded-integer`], and [`intx`] exist, but they come with their own host of problems:
+
+* None of these libraries can easily unify with the existing `uN` and `iN` types.
+* Generally, they require a lot of unsafe code to work.
+* These representations tend to be slower and less-optimized than compiler-generated versions.
+* They still require you to generalise integer types with macros instead of const generics.
```

Comment:
I'm not sure I understand the problem here. If a library provides `struct Int<const BITS: usize>(...);` then code using this library shouldn't need macros to interact with it (except, perhaps, as workaround for current gaps in const generics). The library itself would have a bunch of impls relating its types to the language primitives, which may be generated with macros. But that doesn't seem like such a drastic problem, if it's constrained to the innards of one library, or a few competing libraries.

Reply:
I'm not sure I understand your argument. No matter what, a library solution cannot be both generic *and* unify with the standard library types. I don't see a path forward that would allow, for example, some library `Uint<N>` type to allow `Uint<8>` being an alias for `u8` while also supporting arbitrary `Uint<N>`. Even with specialisation, I can't imagine a sound subset of specialisation allowing this to work.

Like, sure, a set of libraries can choose to only use these types instead of the primitives, circumventing the problem. But most people will want to implement their traits for primitives for interoperability.

Reply:
This overlaps a bit with the bullet point about unification, but I do think it depends a lot on what one is doing. For example, the num-traits crate defines traits that it needs to implement for the primitive types. On the other hand, any code that's currently written against the traits from num-traits may be happy with a third party library that provides `Int<N>` and `Uint<N>` and implements the relevant traits for them. And for something like bit fields, you may not need much interop with the primitive types at all: in the `MipsInstruction` example, you probably want some widening and narrowing conversions, but only with respect to u32 specifically.

It's hard to form an opinion about how common these scenarios are (and whether there are other nuances) without having a corpus of "real" code to look at. Experience reports (including negative ones) with crates like num-traits and bounded-integer may be more useful than discussing it in the abstract.